### ISSUE TITLE
Possible error in delete/registration code

### ISSUE THREAD

**Issue by gbraad** [minishift/minishift#1121]
When deleting a CentOS based image using `minishift delete`, I get the following error:

```
$ minishift delete
Current Minishift VM is registered, but unregistration failed. Do you still want to delete the VM [y/N]?: y
Deleting the Minishift VM...
Minishift VM deleted.
```

This seems wrong, as the VM is not `registered`.

**Comment by praveenkumar** [comment]
@gbraad It might be that you tried CDK first and delete doesn't happen properly so `registration` flag is configuration file still set to `true` which cased this issue.

**Comment by gbraad** [comment]
On this machine I have never used CDK...

**Comment by praveenkumar** [comment]
@gbraad I am sure pretty sure below is situation for you. It not about you tried CDK on that machine but even you used RHEL iso to try out something and then that didn't removed or deleted properly.

```
$ cat ~/.minishift/machines/minishift.json
{
	"OcPath": "/home/prkumar/.minishift/cache/oc/v1.5.1/oc",
	"IsRegistered": true,  ===> This is what showing that message.
	"HostFolders": []
}
```

**Comment by gbraad** [comment]
The problem is, that is the default value due to https://github.com/minishift/minishift/blob/5c02dee1f8814d07dace81a11e594bdfbc79a1db/cmd/minishift/cmd/start.go#L550 Which means that even when NOT using an RHEL image, it will mark the flag as `IsRegistered`: (never used CDK or a RHEL image on this machine)

**Comment by gbraad** [comment]
```
$ minishift delete --clear-cache
$ rm -rf ~/.minishift/machines/  # to be sure
$ vi cmd/minishift/cmd/start.go
```

`cmd/minishift/cmd/start.go`
```go
 541 func registerHost(libMachineClient *libmachine.Client) {
 542         if viper.GetBool(startFlags.SkipRegistration.Name) {
 543                 log.Debug("Skipping registration due to enabled --skip-registration flag")
 544                 return
 545         }
 546         if err := cluster.Register(libMachineClient); err != nil {
 547                 atexit.ExitWithMessage(1, fmt.Sprintf("Error to register VM: %v", err))
 548         } else {
 549                 println("REGISTER ???")
 550                 minishiftConfig.InstanceConfig.IsRegistered = true
 551                 minishiftConfig.InstanceConfig.Write()
 552         }
 553 }
```

```
$ minishift start
Starting local OpenShift cluster using 'kvm' hypervisor...
Downloading ISO 'https://github.com/minishift/minishift-b2d-iso/releases/download/v1.0.2/minishift-b2d.iso'
 40.00 MiB / 40.00 MiB [===========================================================================================================================================================================================] 100.00% 0s
REGISTER ???
-- Checking OpenShift client ... OK
-- Checking Docker client ... OK
-- Checking Docker version ... OK
-- Checking for existing OpenShift container ... OK
-- Checking for openshift/origin:v1.5.1 image ... 
   Pulling image openshift/origin:v1.5.1
   Pulled 0/3 layers, 3% complete
   Pulled 0/3 layers, 32% complete
   Pulled 0/3 layers, 67% complete
```

This means that for any image we use, `IsRegistered` will be set in the `start` command due to the `registerHost` being called:

`cmd/minishift/cmd/start.go`
```go
 139         hostVm, ip := startHost(libMachineClient)
 140         registerHost(libMachineClient)
```

**Comment by praveenkumar** [comment]
@gbraad I will check that, looks like logical issue :(

**Comment by gbraad** [comment]
It feels like a tristate has intended here:

```go
 546         if isNeeded, err := cluster.Register(libMachineClient); isNeeded && err != nil {
 547                 atexit.ExitWithMessage(1, fmt.Sprintf("Error to register VM: %v", err))
 548         } else {  // registration is needed and didn't fail
 549                 minishiftConfig.InstanceConfig.IsRegistered = true
 550                 minishiftConfig.InstanceConfig.Write()
 551         }
```

to handle the three situations:

  * Registration is needed, and
    * failed (show error)
    * succeeded (set `IsRegistered`)
  * Registration is not needed

but to me, instead of having the logic being handled in `start.registerHost()`, it should happen in `cluster.Register()` is this code has all the knowledge to make this decision. This would also simplify the function to just:

```go
 546         if err := cluster.Register(libMachineClient); isNeeded && err != nil {
 547                 atexit.ExitWithMessage(1, fmt.Sprintf("Error to register VM: %v", err))
 548         }
```

as this would imply registration was needed (and not skipped). And instead, `Register` would do something like:

```diff
// Register host VM
@@ -144,6 +144,8 @@ func Register(api libmachine.API) error {
	host, err := api.Load(constants.MachineName)
	if err != nil {
		return err
	}
	if err := registration.RegisterHostVM(host, RegistrationParameters); err != nil {
		return err
	}

+	minishiftConfig.InstanceConfig.IsRegistered = true
+	minishiftConfig.InstanceConfig.Write()
	return nil
}
```
[ref](https://github.com/minishift/minishift/blob/854d63c5d069cddb54b0ae9662938f0ada063b4f/pkg/minikube/cluster/cluster.go#L136)

**Comment by hferentschik** [comment]
Merged via pull request #1140

**Comment by gbraad** [comment]
@hferentschik @praveenkumar @LalatenduMohanty Shall we create a task to handle the technical debt of moving the `Register` and `UnRegister` out of `cluster.go`?
